File: src/main/java/com/joecollins/models/general/Aggregators.java
package com.joecollins.models.general;

import com.joecollins.bindings.Binding;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.function.BinaryOperator;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.ToDoubleFunction;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.apache.commons.lang3.tuple.Pair;
import org.jetbrains.annotations.NotNull;

public class Aggregators {

  private Aggregators() {}

  public static <T, K> Binding<Map<K, Integer>> combine(
      Collection<T> items, Function<T, Binding<Map<K, Integer>>> result) {
    return combine(items, result, new HashMap<>());
  }

  public static <T, K> Binding<Map<K, Integer>> combine(
      Collection<T> items, Function<T, Binding<Map<K, Integer>>> result, Map<K, Integer> identity) {
    Set<K> seededKeys = new HashSet<>(identity.keySet());
    return Binding.mapReduceBinding(
        items.stream().map(result).collect(Collectors.toList()),
        new LinkedHashMap<>(identity),
        (a, r) -> {
          if (r == null) {
            return new LinkedHashMap<>(a);
          }
          return Stream.concat(a.entrySet().stream(), r.entrySet().stream())
              .collect(
                  Collectors.toMap(
                      Map.Entry::getKey, Map.Entry::getValue, Integer::sum, LinkedHashMap::new));
        },
        (a, r) -> {
          if (r == null) {
            return new LinkedHashMap<>(a);
          }
          return Stream.concat(
                  a.entrySet().stream(),
                  r.entrySet().stream().map(e -> ImmutablePair.of(e.getKey(), -e.getValue())))
              .collect(
                  Collectors.toMap(
                      Map.Entry::getKey, Map.Entry::getValue, Integer::sum, LinkedHashMap::new))
              .entrySet()
              .stream()
              .filter(
                  e ->
                      seededKeys.contains(e.getKey())
                          || e.getValue() != 0
                          || !r.containsKey(e.getKey()))
              .collect(
                  Collectors.toMap(
                      Map.Entry::getKey, Map.Entry::getValue, Integer::sum, LinkedHashMap::new));
        });
  }

  public static <T, K> Binding<Map<K, Pair<Integer, Integer>>> combineDual(
      Collection<T> items, Function<T, Binding<Map<K, Pair<Integer, Integer>>>> result) {
    return combineDual(items, result, new HashMap<>());
  }

  public static <T, K> Binding<Map<K, Pair<Integer, Integer>>> combineDual(
      Collection<T> items,
      Function<T, Binding<Map<K, Pair<Integer, Integer>>>> result,
      Map<K, Pair<Integer, Integer>> identity) {
    Set<K> seededKeys = new HashSet<>(identity.keySet());
    BinaryOperator<Pair<Integer, Integer>> sum =
        (v1, v2) -> ImmutablePair.of(v1.getLeft() + v2.getLeft(), v1.getRight() + v2.getRight());
    return Binding.mapReduceBinding(
        items.stream().map(result).collect(Collectors.toList()),
        new LinkedHashMap<>(identity),
        (a, r) -> {
          if (r == null) {
            return new LinkedHashMap<>(a);
          }
          return Stream.concat(a.entrySet().stream(), r.entrySet().stream())
              .collect(
                  Collectors.toMap(
                      Map.Entry::getKey, Map.Entry::getValue, sum, LinkedHashMap::new));
        },
        (a, r) -> {
          if (r == null) {
            return new LinkedHashMap<>(a);
          }
          return Stream.concat(
                  a.entrySet().stream(),
                  r.entrySet().stream()
                      .map(
                          e ->
                              new ImmutablePair<>(
                                  e.getKey(),
                                  ImmutablePair.of(
                                      -e.getValue().getLeft(), -e.getValue().getRight()))))
              .collect(
                  Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, sum, LinkedHashMap::new))
              .entrySet()
              .stream()
              .filter(
                  e ->
                      seededKeys.contains(e.getKey())
                          || e.getValue().getLeft() != 0
                          || e.getValue().getRight() != 0
                          || !r.containsKey(e.getKey()))
              .collect(
                  Collectors.toMap(
                      Map.Entry::getKey, Map.Entry::getValue, sum, LinkedHashMap::new));
        });
  }

  public static <T> Binding<Integer> sum(Collection<T> items, Function<T, Binding<Integer>> val) {
    return Binding.mapReduceBinding(
        items.stream().map(val).collect(Collectors.toList()), 0, (t, v) -> t + v, (t, v) -> t - v);
  }

  public static <K> Binding<Map<K, Integer>> adjustForPctReporting(
      Binding<Map<K, Integer>> result, Binding<Double> pctReporting) {
    return result.merge(
        pctReporting,
        (r, p) -> {
          LinkedHashMap<K, Integer> ret = new LinkedHashMap<>();
          r.forEach((k, v) -> ret.put(k, (int) (v * p)));
          return ret;
        });
  }

  public static <T> Binding<Double> combinePctReporting(
      Collection<T> items, Function<T, Binding<Double>> pctReportingFunc) {
    return combinePctReporting(items, pctReportingFunc, t -> 1);
  }

  public static <T> Binding<Double> combinePctReporting(
      Collection<T> items,
      Function<T, Binding<Double>> pctReportingFunc,
      ToDoubleFunction<T> weightFunc) {
    double totalWeight = items.stream().mapToDouble(weightFunc).sum();
    return Binding.mapReduceBinding(
        items.stream()
            .map(
                t -> {
                  double weight = weightFunc.applyAsDouble(t);
                  return pctReportingFunc.apply(t).map(x -> x * weight);
                })
            .collect(Collectors.toList()),
        0.0,
        (a, p) -> a + (p / totalWeight),
        (a, p) -> a - (p / totalWeight));
  }

  public static <K1, K2> Binding<Map<K2, Integer>> adjustKey(
      Binding<Map<K1, Integer>> result, Function<K1, K2> func) {
    return result.map(m -> adjustKey(m, func));
  }

  public static <K1, K2> Map<K2, Integer> adjustKey(Map<K1, Integer> map, Function<K1, K2> func) {
    LinkedHashMap<K2, Integer> ret = new LinkedHashMap<>();
    map.forEach((k, v) -> ret.merge(func.apply(k), v, Integer::sum));
    return ret;
  }

  public static <K> Binding<Map<K, Double>> toPct(Binding<Map<K, Integer>> mapBinding) {
    return mapBinding.map(Aggregators::toPct);
  }

  public static <K> Map<K, Double> toPct(Map<K, Integer> map) {
    int total = map.values().stream().mapToInt(i -> i).sum();
    return map.entrySet().stream()
        .collect(
            Collectors.toMap(
                Map.Entry::getKey,
                e -> total == 0 ? 0 : 1.0 * e.getValue() / total,
                Double::sum,
                LinkedHashMap::new));
  }

  public static <K> Binding<Map<K, Integer>> topAndOthers(
      Binding<? extends Map<K, Integer>> result, int limit, K others) {
    return topAndOthers(result, limit, others, Binding.fixedBinding((K[]) new Object[0]));
  }

  public static <K> Binding<Map<K, Integer>> topAndOthers(
      Binding<? extends Map<K, Integer>> result, int limit, K others, Binding<K[]> mustInclude) {
    return result.merge(mustInclude, (m, w) -> topAndOthers(m, limit, others, w));
  }

  @SafeVarargs
  public static <K> Map<K, Integer> topAndOthers(
      Map<K, Integer> map, int limit, K others, K... mustInclude) {
    if (map.size() <= limit) {
      return map;
    }
    Set<K> mustIncludeSet =
        Arrays.stream(mustInclude).filter(Objects::nonNull).collect(Collectors.toSet());
    Set<Map.Entry<K, Integer>> top =
        map.entrySet().stream()
            .filter(e -> !mustIncludeSet.contains(e.getKey()))
            .sorted(
                Map.Entry.<K, Integer>comparingByValue(
                        Comparator.comparingInt(i -> i == null ? -1 : i))
                    .reversed())
            .limit(Math.max(0, limit - 1 - mustIncludeSet.size()))
            .collect(Collectors.toSet());
    Predicate<Map.Entry<K, Integer>> topAndRequired =
        e -> top.contains(e) || mustIncludeSet.contains(e.getKey());
    LinkedHashMap<K, Integer> result = new LinkedHashMap<>();
    boolean needOthers = false;
    for (Map.Entry<K, Integer> e : map.entrySet()) {
      if (topAndRequired.test(e) || e.getValue() != null) {
        var key = topAndRequired.test(e) ? e.getKey() : others;
        if (result.containsKey(key)) {
          result.merge(key, e.getValue(), Integer::sum);
        } else {
          result.put(key, e.getValue());
        }
      } else {
        needOthers = true;
      }
    }
    if (needOthers) {
      result.put(others, null);
    }
    return result;
  }

  public static <K, V> Binding<Map<K, V>> toMap(
      Collection<K> keys, Function<K, Binding<V>> bindingFunc) {
    return toMap(keys, Function.identity(), bindingFunc);
  }

  public static <T, K, V> Binding<Map<K, V>> toMap(
      Collection<T> entries, final Function<T, K> keyFunc, Function<T, Binding<V>> bindingFunc) {
    return new Binding<>() {
      private List<Binding<V>> bindings = null;
      private Map<K, V> value = null;

      @Override
      public Map<K, V> getValue() {
        return entries.stream()
            .collect(Collectors.toMap(keyFunc, k -> bindingFunc.apply(k).getValue()));
      }

      @Override
      public void bind(@NotNull Function1<? super Map<K, V>, Unit> onUpdate) {
        bindLegacy((Consumer<Map<K, V>>) onUpdate::invoke);
      }

      @Override
      public void bindLegacy(Consumer<Map<K, V>> onUpdate) {
        if (bindings != null) {
          throw new IllegalStateException("Binding is already used");
        }
        value = new HashMap<>();
        Map<K, Binding<V>> bindingsMap = new HashMap<>();
        for (T entry : entries) {
          K key = keyFunc.apply(entry);
          Binding<V> binding = bindingFunc.apply(entry);
          bindingsMap.put(key, binding);
          binding.bindLegacy(
              val -> {
                value.put(key, val);
                if (bindings != null) {
                  onUpdate.accept(value);
                }
              });
        }
        onUpdate.accept(value);
        bindings = new ArrayList<>(bindingsMap.values());
      }

      @Override
      public void unbind() {
        if (bindings != null) {
          bindings.forEach(Binding::unbind);
          bindings = null;
          value = null;
        }
      }
    };
  }
}

File: src/main/java/com/joecollins/models/general/Aggregators.kt
package com.joecollins.models.general

import com.joecollins.bindings.Binding
import java.lang.Integer.max
import org.apache.commons.lang3.tuple.ImmutablePair
import org.apache.commons.lang3.tuple.Pair

object Aggregators {

    @JvmStatic fun <T, K> combine(items: Collection<T>, result: (T) -> Binding<Map<K, Int>>) = combine(items, result, HashMap())

    @JvmStatic fun <T, K> combine(items: Collection<T>, result: (T) -> Binding<Map<K, Int>>, identity: Map<K, Int> = HashMap()): Binding<Map<K, Int>> {
        val seededKeys = identity.keys
        return Binding.mapReduceBinding(
            items.map(result),
            identity,
            { a, r ->
                val ret = LinkedHashMap(a)
                r.forEach { (k, v) -> ret.merge(k, v) { a, b -> a + b } }
                ret
            },
            { a, r ->
                val ret = LinkedHashMap(a)
                r.forEach { (k, v) -> ret.merge(k, -v) { a, b -> a + b } }
                ret.filter { seededKeys.contains(it.key) || it.value != 0 || !r.containsKey(it.key) }
            }
        )
    }

    @JvmStatic fun <T, K> combineDual(items: Collection<T>, result: (T) -> Binding<Map<K, Pair<Int, Int>>>) = combineDual(items, result, HashMap())

    @JvmStatic fun <T, K> combineDual(items: Collection<T>, result: (T) -> Binding<Map<K, Pair<Int, Int>>>, identity: Map<K, Pair<Int, Int>> = HashMap()): Binding<Map<K, Pair<Int, Int>>> {
        val seededKeys = identity.keys
        val sum: (Pair<Int, Int>, Pair<Int, Int>) -> Pair<Int, Int> =
            { a, b -> ImmutablePair.of(a.left + b.left, a.right + b.right) }
        return Binding.mapReduceBinding(
            items.map(result),
            identity,
            { a, r ->
                val ret = LinkedHashMap(a)
                r.forEach { (k, v) -> ret.merge(k, v, sum) }
                ret
            },
            { a, r ->
                val ret = LinkedHashMap(a)
                r.forEach { (k, v) -> ret.merge(k, ImmutablePair.of(-v.left, -v.right), sum) }
                ret.filter { e -> seededKeys.contains(e.key) || e.value.left != 0 || e.value.right != 0 || !r.containsKey(e.key) }
            }
        )
    }

    @JvmStatic fun <T> sum(items: Collection<T>, value: (T) -> Binding<Int>): Binding<Int> {
        return Binding.mapReduceBinding(
            items.map(value), 0, { t, v -> t + v }, { t, v -> t - v }
        )
    }

    @JvmStatic fun <K> adjustForPctReporting(result: Binding<Map<K, Int>>, pctReporting: Binding<Double>): Binding<Map<K, Int>> {
        return result.merge(pctReporting) { r, p ->
            val ret: LinkedHashMap<K, Int> = LinkedHashMap()
            r.forEach { (k, v) -> ret[k] = (v * p).toInt() }
            ret
        }
    }

    @JvmStatic fun <T> combinePctReporting(items: Collection<T>, pctReportingFunc: (T) -> Binding<Double>) = combinePctReporting(items, pctReportingFunc) { 1.0 }

    @JvmStatic fun <T> combinePctReporting(items: Collection<T>, pctReportingFunc: (T) -> Binding<Double>, weightFunc: (T) -> Double): Binding<Double> {
        val totalWeight = items.map(weightFunc).sum()
        return Binding.mapReduceBinding(
            items.map { e ->
                val weight = weightFunc(e)
                pctReportingFunc(e).map { it * weight } },
            0.0,
            { a, p -> a + (p / totalWeight) },
            { a, p -> a - (p / totalWeight) }
        )
    }

    @JvmStatic fun <K1, K2> adjustKey(result: Binding<Map<K1, Int>>, func: (K1) -> K2) = result.map { adjustKey(it, func) }

    @JvmStatic fun <K1, K2> adjustKey(result: Map<K1, Int>, func: (K1) -> K2): Map<K2, Int> {
        val ret: LinkedHashMap<K2, Int> = LinkedHashMap()
        result.forEach { (k, v) -> ret.merge(func(k), v) { a, b -> a + b } }
        return ret
    }

    @JvmStatic fun <K> toPct(result: Binding<Map<K, Int>>) = result.map { toPct(it) }

    @JvmStatic fun <K> toPct(result: Map<K, Int>): Map<K, Double> {
        val total = result.values.sum()
        return result.mapValues { if (total == 0) 0.0 else (1.0 * it.value / total) }
    }

    @Suppress("UNCHECKED_CAST")
    @JvmStatic fun <K> topAndOthers(result: Binding<Map<K, Int>>, limit: Int, others: K) = topAndOthers(result, limit, others, Binding.fixedBinding(Array<Any?>(0) { null } as Array<K>))

    @JvmStatic fun <K> topAndOthers(result: Binding<Map<K, Int>>, limit: Int, others: K, mustInclude: Binding<Array<K>>) = result.merge(mustInclude) { m, w -> topAndOthers(m, limit, others, *w) }

    @Suppress("UNCHECKED_CAST")
    @JvmStatic fun <K, T : Int?> topAndOthers(result: Map<K, T>, limit: Int, others: K, vararg mustInclude: K): Map<K, T> {
        if (result.size <= limit) {
            return result
        }
        val mustIncludeSet = mustInclude.filterNotNull().toSet()
        val top = result.entries
            .filter { !mustIncludeSet.contains(it.key) }
            .sortedByDescending { it.value ?: -1 }
            .take(max(0, limit - 1 - mustIncludeSet.size))
            .toSet()
        val topAndRequired: (Map.Entry<K, Int?>) -> Boolean = { top.contains(it) || mustIncludeSet.contains(it.key) }
        var needOthers = false
        val ret: LinkedHashMap<K, T> = LinkedHashMap()
        result.entries.forEach { e ->
            if (topAndRequired(e) || e.value != null) {
                val key = if (topAndRequired(e)) e.key else others
                if (ret.containsKey(key)) ret.merge(key, e.value!!) { a, b -> (a!!.toInt() + b!!.toInt()) as T }
                else ret[key] = e.value
            } else {
                needOthers = true
            }
        }
        if (needOthers) {
            ret[others] = null as T
        }
        return ret
    }

    @JvmStatic fun <K, V> toMap(keys: Collection<K>, bindingFunc: (K) -> Binding<V>) = toMap(keys, { it }, bindingFunc)

    @JvmStatic fun <T, K, V> toMap(entries: Collection<T>, keyFunc: (T) -> K, bindingFunc: (T) -> Binding<V>): Binding<Map<K, V>> {
        return object : Binding<Map<K, V>> {
            private var bindings: List<Binding<V>>? = null
            private var map: Map<K, V>? = null

            override val value get() = entries.map { keyFunc(it) to bindingFunc(it).value }.toMap()
            override fun bind(onUpdate: (Map<K, V>) -> Unit) {
                check(bindings == null) { "Binding is already used" }
                map = HashMap()
                val bindingsMap = HashMap<K, Binding<V>>()
                entries.forEach { e ->
                    val key = keyFunc(e)
                    val binding = bindingFunc(e)
                    bindingsMap[key] = binding
                    binding.bind {
                        val updated = HashMap<K, V>(map)
                        updated[key] = it
                        map = updated
                        if (bindings != null) onUpdate(updated)
                    }
                }
                onUpdate(map!!)
                bindings = bindingsMap.values.toList()
            }
            override fun unbind() {
                val bindings = this.bindings
                if (bindings != null) {
                    bindings.forEach { it.unbind() }
                    this.bindings = null
                    map = null
                }
            }
        }
    }
}
